File: src/01_data_structures/04_stack/implementations/ts/stack.ts
interface IStack<T> {
  push(value: any): IStack<T>;
  pop(): T | null;
  length(): number;
  isEmpty(): boolean;
  peek(): T | null;
}

export class Stack<T> implements IStack<T> {
  private storage: Array<T> = [];

  push(value: any): Stack<T> {
    this.storage.push(value);

    return this;
  }

  pop(): T | null {
    if (this.storage.length === 0) {
      return null;
    }

    return this.storage.pop();
  }

  length(): number {
    return this.storage.length;
  }

  isEmpty(): boolean {
    return this.length() === 0;
  }

  peek(): T | null {
    if (this.isEmpty()) {
      return null;
    }

    return this.storage[this.storage.length - 1];
  }
}

File: src/01_data_structures/04_stack/implementations/rs/stack.rs
pub struct Stack<T> {
  storage: Vec<T>,
}

impl<T> Stack<T> {
  pub fn new() -> Self {
    Stack { storage: Vec::new() }
  }

  pub fn push(mut self, value: T) -> Self {
    self.storage.push(value);

    return self;
  }

  pub fn pop(&mut self) -> Option<T> {
    if self.storage.len() == 0 {
      return None;
    }

    self.storage.pop()
  }

  pub fn length(&self) -> usize {
    self.storage.len()
  }

  pub fn is_empty(&self) -> bool {
    self.storage.len() == 0
  }

  pub fn peek(&self) -> Option<&T> {
    self.storage.last()
  }
}
